### PR TITLE
Performance: Optimize the AddTemplate component used in data views pages

### PR DESCRIPTION
**Pull request by youknowriad**
## What?

I was debugging the templates page and noticed that the "Add Template" component re-renders way too often. For instance, why is it re-rendering when I'm filtering the table.

I believe we can do better than this PR but at least this is a first step.

## Testing Instructions

1- Open the templates page in the site editor
2- You can add a "console.log" to the AddTemplate component.
3- Notice that as you search, the component doesn't re-render.

### CODE REVIEW

File: packages/edit-site/src/components/add-new-template/index.js [hunk 1]
```diff
@@ -1,31 +1,431 @@
+/**
```
Comment (youknowriad) [inline packages/edit-site/src/components/add-new-template/index.js:1]: I've moved the old new-template file into this index file. There's no need to separate these two things and there's no need for props since this component is now used only for templates and only in the data views page.

File: packages/edit-site/src/components/add-new-template/index.js [hunk 2]
```diff
@@ -1,31 +1,431 @@
...
+}
 
-	if ( ! postType ) {
+const modalContentMap = {
+	templatesList: 1,
+	customTemplate: 2,
+	customGenericTemplate: 3,
+};
+
+function NewTemplate() {
```
Comment (youknowriad) [inline packages/edit-site/src/components/add-new-template/index.js]: I believe we should split the modal into its own component. But it can be done separately.

File: packages/edit-site/src/components/add-new-template/utils.js [hunk 3]
```diff
@@ -622,41 +624,57 @@ const useTemplatesToExclude = (
 const useEntitiesInfo = (
 	entityName,
 	templatePrefixes,
-	additionalQueryParameters = {}
+	additionalQueryParameters = EMPTY_OBJECT
 ) => {
```
Comment (youknowriad) [inline packages/edit-site/src/components/add-new-template/utils.js:628]: This was throwing a React warning because the useSelect was re-rendering the components while the data didn't change.

I just refactored it a bit to solve the above, but to be honest these hooks are very unclear, I'm not even sure we need them and why they exist as helpers. They just make things hard to follow.

File: packages/edit-site/src/components/add-new-template/index.js (hunk 2 above)
Comment (youknowriad) [reply packages/edit-site/src/components/add-new-template/index.js]: Ok I just went ahead and did this.

File: packages/edit-site/src/components/page-templates-template-parts/add-new-template-part.js [hunk 4]
```diff
@@ -58,3 +58,5 @@ export default function AddNewTemplatePart() {
 		</>
 	);
 }
+
+export default memo( AddNewTemplatePart );
```
Comment (ellatrix) [inline packages/edit-site/src/components/page-templates-template-parts/add-new-template-part.js:62]: Was there anything in this component that made the re-renders expensive?

Comment (youknowriad) [reply packages/edit-site/src/components/page-templates-template-parts/add-new-template-part.js:62]: I actually was testing "templates" mainly, so this one is probably less expensive but I just thought that it doesn't really make sense to re-render this one when you're interacting with the data views. (like filtering...)

**Review by ellatrix — APPROVED** [review]

**Comment by youknowriad** [comment]
I think this PR improves the "navigate" metric (hard to tell on a single run because that metric is a bit unstable)
